### PR TITLE
Validate and throws a custom error when attempting to register invalid hook functions

### PR DESCRIPTION
**Pull request by jhhom**
Handles [this issue](https://github.com/fastify/fastify/issues/4320), a hook is invalid if it is not of type function because it needs to bind with the fastify instance.

This PR will add validation for the hook functions passed in the `options` argument of [route registration functions](https://www.fastify.io/docs/latest/Reference/Routes/#shorthand-declaration).

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by cesarvspr — APPROVED** [review]
lgtm

File: lib/route.js [hunk 1]
```diff
@@ -310,7 +310,12 @@ function buildRouting (options) {
           for (const hook of lifecycleHooks) {
             const toSet = this[kHooks][hook]
               .concat(opts[hook] || [])
-              .map(h => h.bind(this))
+              .map(h => {
+                if (typeof h !== 'function') {
```
Comment (climba03003) [inline lib/route.js:314]: Can you move the logic right after the `onRoute` hook execution.
In here, the `typeof` validation will run through again and again for the same function.

We only need to guarantee the newly added `function` to be valid.

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: lib/route.js (hunk 1 above)
Comment (jhhom) [reply lib/route.js]: Got it, I have moved the logic to right after the `onRoute` hook execution in my latest commit.

Also added new test to pass the coverage check.

File: lib/route.js [hunk 2]
```diff
@@ -243,6 +243,14 @@ function buildRouting (options) {
         }
       }
 
+      for (const hook of lifecycleHooks) {
+        [].concat(opts[hook] || []).forEach(h => {
+          if (typeof h !== 'function') {
+            throw new Error(`${hook} hook should be a function, instead got ${typeof h}`)
+          }
+        })
```
Comment (climba03003) [inline lib/route.js:251]:
```suggestion
        if(hook in opts) {
          for(const func of opts[hook]) {
            if(typeof func !== 'function') {
              throw new Error(`${hook} hook should be a function, instead got ${typeof func}`)
            }
          }
        }
```
```

**Review by climba03003 — APPROVED** [review]

**Review by cesarvspr — APPROVED** [review]

File: test/hooks.test.js [hunk 3]
```diff
@@ -3301,3 +3301,31 @@ test('onTimeout should be triggered and socket _meta is set', t => {
     })
   })
 })
+
+test('registering invalid hooks should throw an error', async t => {
```
Comment (Eomm) [inline test/hooks.test.js:3305]: Could you add a test where an `onRoute` hook adds an `undefined` hook?

Here is an example code: https://www.fastify.io/docs/latest/Guides/Plugins-Guide/#hooks

Comment (jhhom) [reply test/hooks.test.js:3305]: Sure, added the test in my latest commit.

**Review by Eomm — APPROVED** [review]

**Review by delvedor — CHANGES_REQUESTED** [review]
Thanks for the pr!
Can you use/create a custom error like we are doing for every other Fastify custom error? See [here](https://github.com/fastify/fastify/blob/main/lib/errors.js).
Thanks!

**Review by delvedor — CHANGES_REQUESTED** [review]
It seems that we have this error already: https://github.com/fastify/fastify/blob/acba25acc3e97e3598daf79537a39a5272dba7b8/lib/errors.js?rgh-link-date=2022-10-13T09%3A27%3A23Z#L98-L103

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by jhhom** [comment]
@delvedor I think the error message I added (e.g "onSend hook should be a function, instead got undefined") can be more informative than the error message in the existing error "The hook callback must be a function", it also provides the following additional information which may helps in debugging:

* which hook is invalid
* the actual type passed

In my latest commit, I modified the error message of the existing error instead.
Just that I'm not sure if modifying the error message of the existing error might have any backwards-compatibility issues. I think its probably not an issue because error checking should be done using the error code.

**Comment by delvedor** [comment]
I see your point, then I'll suggest to update the existing error message :)

> Just that I'm not sure if modifying the error message of the existing error might have any backwards-compatibility issues. I think its probably not an issue because error checking should be done using the error code.

This is a valid point, but we have error codes for a reason. Error messages are the human-friendly version of them which is subjected to change.

**Comment by jhhom** [comment]
> I see your point, then I'll suggest to update the existing error message :)

the existing error message is updated in my last commit 👍

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
@delvedor PTAL

**Review by Uzlopak — APPROVED** [review]

**Review by cesarvspr — APPROVED** [review]
Nice

**Review by delvedor — APPROVED** [review]
LGTM

**Pull request merged**